File: src/main/java/com/plugin/frege/psi/FregeElementTypeFactory.java
package com.plugin.frege.psi;

import com.intellij.psi.tree.IElementType;
import com.plugin.frege.stubs.types.FregeDataNativeNameElementType;
import com.plugin.frege.stubs.types.FregeFunctionNameElementType;
import com.plugin.frege.stubs.types.FregePackageClassNameElementType;

public class FregeElementTypeFactory {
    public static IElementType factory(String name) {
        switch (name) {
            case "DATA_NAME_NATIVE":
                return new FregeDataNativeNameElementType(name);
            case "PACKAGE_CLASS_NAME":
                return new FregePackageClassNameElementType(name);
            case "FUNCTION_NAME":
                return new FregeFunctionNameElementType(name);
            default:
                throw new IllegalStateException("Unknown element name: " + name);
        }
    }
}

File: src/main/kotlin/com/plugin/frege/psi/FregeElementTypeFactory.kt
package com.plugin.frege.psi

import com.intellij.psi.tree.IElementType
import com.plugin.frege.stubs.types.*

object FregeElementTypeFactory {
    @JvmStatic
    fun factory(name: String): IElementType {
        return when (name) {
            "NATIVE_DATA_DECL" -> FregeNativeDataDeclElementType(name)
            "PROGRAM" -> FregeProgramElementType(name)
            "CLASS_DECL" -> FregeClassDeclElementType(name)
            "DATA_DECL" -> FregeDataDeclElementType(name)
            "NEWTYPE_DECL" -> FregeNewtypeDeclElementType(name)
            "TYPE_DECL" -> FregeTypeDeclElementType(name)
            "CONSTRUCT" -> FregeConstructElementType(name)
            "BINDING" -> FregeBindingElementType(name)
            "ANNOTATION_ITEM" -> FregeAnnotationItemElementType(name)
            "NATIVE_FUNCTION" -> FregeNativeFunctionElementType(name)
            else -> throw IllegalStateException("Unknown element name: $name")
        }
    }
}
